extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/runtime/Mutex.java

```diff
@@ -0,0 +1,97 @@
+package io.quarkus.websockets.next.runtime;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.websockets.next.WebSocketServerConnection;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+
+/**
+ * TODO implement a more robust solution to ensure that endpoint callbacks are not invoked concurrently.
+ */
+class Mutex {
+
+    private static final Logger LOG = Logger.getLogger(Mutex.class);
+
+    private final Context context;
+    private final WebSocketServerConnection connection;
+
+    private volatile boolean acquired;
+    private final Lock lock = new ReentrantLock();
+    private final ConcurrentLinkedQueue<Runnable> queue;
+
+    Mutex(Context context, WebSocketServerConnection connection) {
+        this.context = context;
+        this.connection = connection;
+        this.queue = new ConcurrentLinkedQueue<>();
+    }
+
+    Consumer<Void> promiseComplete(Promise<Void> promise) {
+        return new PromiseComplete(promise);
+    }
+
+    void run(Runnable action) {
+        boolean run = false;
```

Comment:
yep, you can use a write combiner eg

```java

private final Queue q = // some mpsc unbounded JCtools queue which we already expose via mutiny/netty
private final AtomicLong inFlights;

void run(Runnable action) {
      if (inflights.compareAndSet(0, 1) {
            action.run();            
      } else {
            q.add(action);
            // we need to make sure that at least one completion is in flight, to keep on draining, in case
            if (inflights.getAndIncrement() == 0) {
                 Runnable runnable = q.poll();
                 runnable.run();
            }
      }
}


class PromiseComplete implement Consumer<Void> {
     
        final Promise<Void> promise;

        private PromiseComplete(Promise<Void> promise) {
            this.promise = promise;
        }

        @Override
        public void accept(Void t) {
            try {
                promise.complete();
            } finally {
                // this is expecting its related Runnable::run to be completed, either while run on Mutex::run or while executed from the context
                if (inflights.decrementAndGet() == 0) {
                    // we've done here
                    return;       
                }
                Runnable action = queue.poll();
                assert action != null;
                context.runOnContext(new Handler<Void>() {
                        @Override
                        public void handle(Void event) {
                            action.run();                              
                        }
                });
            }
        }
}



}
```

this is using a variation of https://akarnokd.blogspot.com/2015/05/operator-concurrency-primitives_11.html
to keep a reactive-like ordered execution of events, optimizing for the non-contended case

Reply:
I've tried to implement something like this ^. And renamed the `Mutex` class to `Bulkhead` ;-).

Reply:
That is not a bulkhead, that's just a plain old queue. It does not _limit_ concurrent invocations; it _serializes_ them.

Reply:
It's a bulkhead that limits concurrent invocations to `1` (like `@Bulkhead(1)` with MP Fault Tolerance ;-). It's not a queue as if no concurrent invocations occur then the action is not queued up at all but executed right away. Maybe we could call is `Serializer` but that's IMO a too overloaded term. In any case, it's an internal package-private class so the name does not really matter...

Reply:
Nah, a `@Bulkhead(1)` will actually throw an exception if you try to execute a 2nd task :-) If your tasks are asynchronous, a MP FT `@Bulkhead` has a ... wait for it ... queue. Which is bounded, by default to 10 tasks, then it will start failing the incoming tasks again.

I agree `Serializer` is a bad name, but a `SerializingExecutor` should be reasonable.

(I actually don't like the name `Bulkhead` as used in MP FT. It is in fact a `ConcurrencyLimiter`. The generic term "bulkhead" refers to system compartmentalization.)

Reply:
In Vertx I propose stage same thing (actually not the same, because can be more batchy, doesn't have any 2-staged execution and is not optimized for non-concurrent use cases): https://github.com/eclipse-vertx/vert.x/blob/a9f70c023be41f8c1621d8384b4f98af6a76099f/src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java#L28

Reply:
> Nah, a @Bulkhead(1) will actually throw an exception if you try to execute a 2nd task :-) If your tasks are asynchronous, a MP FT @Bulkhead has a ... wait for it ... queue. Which is bounded, by default to 10 tasks, then it will start failing the incoming tasks again.

Ofc, I was talking about the async use case as we're async too (i.e. something like `@Bulkhead(1) @Asynchronous` and with a bigger waiting queue).

> I agree Serializer is a bad name, but a SerializingExecutor should be reasonable.

Ok, let's rename it again. But I don't like the name `Executor` because like Franz menioned above this is a two-step execution. First you need to obtain a callback and then you attempt to execute the action.

> (I actually don't like the name Bulkhead as used in MP FT. It is in fact a ConcurrencyLimiter. The generic term "bulkhead" refers to system compartmentalization.)

Fair enough. Let's use `ConcurrencyLimiter` then?